KAFKA-9931: Implement KIP-605 to expand support for Connect worker internal topic configurations

rhauch opened the pull request:
**[KIP-605](https://cwiki.apache.org/confluence/display/KAFKA/KIP-605%3A+Expand+Connect+Worker+Internal+Topic+Settings) has passed.**

Expanded the allowed values for the internal topics’ replication factor and partitions from positive values to also include -1 to signify that the broker defaults should be used.

The Kafka storage classes were already constructing a `NewTopic` object (always with a replication factor and partitions) and sending it to Kafka when required. This change will avoid setting the replication factor and/or number of partitions on this `NewTopic` if the worker configuration uses -1 for the corresponding configuration value.

Quite a few new tests were added to verify that the `TopicAdmin` utility class is correctly using the AdminClient, that the `DistributedConfig` validators for these configurations are correct, and that the `DistributedConfig` is correctly assembling the configuration properties that define the topic settings, which are now accessed by the three Kafka storage objects before they create the topics.

Also added support for additional topic settings used when creating the config, status, and offset internal topics. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-9931: Added support for -1 replication factor and partitions for distributed worker internal topics

Expanded the allowed values for the internal topics’ replication factor and partitions from positive values to also include -1 to signify that the broker defaults should be used.

The Kafka storage classes were already constructing a `NewTopic` object (always with a replication factor and partitions) and sending it to Kafka when required. This change will avoid setting the replication factor and/or number of partitions on this `NewTopic` if the worker configuration uses -1 for the corresponding configuration value.

Quite a few new tests were added to verify that the `TopicAdmin` utility class is correctly using the AdminClient, and that the `DistributedConfig` validators for these configurations are correct.

rhauch committed:
KAFKA-9931: Added support for extra settings for internal topics on distributed config

Added support for additional topic settings used when creating the config, status, and offset internal topics.

rhauch commented:
This includes the commit from #8653, which corrects the replication factor properties in one of the MirrorMaker2 integration test, which should have been:
```
config.storage.replication.factor=1
offset.storage.replication.factor=1
status.storage.replication.factor=1
```
but instead were
```
config.storage.topic.replication.factor=1
offset.storage.topic.replication.factor=1
status.storage.topic.replication.factor=1
```
Note the extra `topic.` in the property name. Because these didn't match the expected `*.storage.replication.factor` properties, they new changes considered the `topic.replication.factor` to be a topic setting that was passed via the `NewTopic` object and `AdminClient` to the broker with the topic creation request. However, that failed with the following exception:
```
[2020-05-11 15:12:01,518] ERROR [Worker clientId=connect-1, groupId=backup-mm2] Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder:297)
org.apache.kafka.connect.errors.ConnectException: Error while attempting to create/find topic(s) 'mm2-offsets.backup.internal'
        at org.apache.kafka.connect.util.TopicAdmin.createTopics(TopicAdmin.java:309)
        at org.apache.kafka.connect.storage.KafkaOffsetBackingStore$1.run(KafkaOffsetBackingStore.java:105)
        at org.apache.kafka.connect.util.KafkaBasedLog.start(KafkaBasedLog.java:128)
        at org.apache.kafka.connect.storage.KafkaOffsetBackingStore.start(KafkaOffsetBackingStore.java:115)
        at org.apache.kafka.connect.runtime.Worker.start(Worker.java:186)
        at org.apache.kafka.connect.runtime.AbstractHerder.startServices(AbstractHerder.java:123)
        at org.apache.kafka.connect.runtime.distributed.DistributedHerder.run(DistributedHerder.java:284)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidConfigurationException: Unknown topic config name: topic.replication.factor
        at org.apache.kafka.common.internals.KafkaFutureImpl.wrapAndThrow(KafkaFutureImpl.java:45)
        at org.apache.kafka.common.internals.KafkaFutureImpl.access$000(KafkaFutureImpl.java:32)
        at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:89)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:260)
        at org.apache.kafka.connect.util.TopicAdmin.createTopics(TopicAdmin.java:274)
        ... 11 more
Caused by: org.apache.kafka.common.errors.InvalidConfigurationException: Unknown topic config name: topic.replication.factor
```

IOW, this change causes a Connect distributed worker or MirrorMaker2 to fail upon startup where it didn't prior to this change if/when the configuration contains extra properties that begin with any of the following prefixes and where the unprefixed configuration does not match a valid topic setting in the Kafka broker being used:
* `offset.storage.*`
* `config.storage.*`
* `status.storage.*`

rhauch commented:
BTW, I've improved the error message when Connect's `TopicAdmin` fails to create a topic because of an unknown topic setting:

> Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder:297)
org.apache.kafka.connect.errors.ConnectException: Unable to create topic(s) 'mm2-offsets.backup.internal': Unknown topic config name: topic.replication.factor


Here's what this looks like in the log message just before the herder exits:
```
[2020-05-14 09:36:22,348] ERROR [Worker clientId=connect-2, groupId=backup-mm2] Uncaught exception in herder work thread, exiting:  (org.apache.kafka.connect.runtime.distributed.DistributedHerder:297)
org.apache.kafka.connect.errors.ConnectException: Unable to create topic(s) 'mm2-offsets.backup.internal': Unknown topic config name: topic.replication.factor
	at org.apache.kafka.connect.util.TopicAdmin.createTopics(TopicAdmin.java:305)
	at org.apache.kafka.connect.storage.KafkaOffsetBackingStore$1.run(KafkaOffsetBackingStore.java:105)
	at org.apache.kafka.connect.util.KafkaBasedLog.start(KafkaBasedLog.java:128)
	at org.apache.kafka.connect.storage.KafkaOffsetBackingStore.start(KafkaOffsetBackingStore.java:115)
	at org.apache.kafka.connect.runtime.Worker.start(Worker.java:186)
	at org.apache.kafka.connect.runtime.AbstractHerder.startServices(AbstractHerder.java:123)
	at org.apache.kafka.connect.runtime.distributed.DistributedHerder.run(DistributedHerder.java:284)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.kafka.common.errors.InvalidConfigurationException: Unknown topic config name: topic.replication.factor
```

rhauch committed:
KAFKA-9931: Added integration tests for internal topic creation

rhauch commented:
I've added some integration tests for creating the internal topic, including verifying some existing functionality w/r/t the replication factor and number of partitions with various Kafka cluster sizes.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -66,25 +71,55 @@
          * @return this builder to allow methods to be chained; never null
          */
         public NewTopicBuilder partitions(int numPartitions) {
+            if (numPartitions == NO_PARTITIONS) {
```
this check seems to replicate a bit what's happening in `NewTopic`. Should we just pass the value, given that we have validations too already?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -66,25 +71,55 @@
          * @return this builder to allow methods to be chained; never null
          */
         public NewTopicBuilder replicationFactor(short replicationFactor) {
+            if (replicationFactor == NO_REPLICATION_FACTOR) {
```
same question as above

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/InternalTopicsIntegrationTest.java:
```diff
@@ -0,0 +1,187 @@
+        connect.removeWorker();
+
+        // Sleep for a bit
+        Thread.sleep(3000);
```
did you notice that this matters? Not suggesting to remove necessarily, but I wonder whether you noticed it didn't work otherwise.

rhauch replied:
Good catch. I had put this in while debugging an issue, but it is not needed.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/InternalTopicsIntegrationTest.java:
```diff
@@ -0,0 +1,187 @@
+    public void testCreateInternalTopicsWithDefaultSettings() throws InterruptedException {
+        int numWorkers = 1;
+        int numBrokers = 3;
+        connect = new EmbeddedConnectCluster.Builder().name("connect-cluster-1")
```
Is naming here to distinguish between runs of integration tests?
I have a commit that will help us know the boundaries between ITs. I'll submit separately.

rhauch replied:
I was running into other issues, and I never changed it. The names are not important, so I'll change them back to the same cluster name.

I actually do prefer how each test fully sets up the cluster builder, in part because some of the information is changed and some isn't. I found it very error prone while I was working on this to reuse a builder and have each test only change *some* of the attributes, or to remember to set the worker props after changing the `workerProps` map.

kkonstantine replied:
Makes sense.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedConfigTest.java:
```diff
@@ -105,4 +106,204 @@ public void shouldValidateAllVerificationAlgorithms() {
+        topicSettings.put("partitions", "3");
+
+        Map<String, String> settings = configs();
+        topicSettings.entrySet().forEach(e -> {
```
`topicSettings.forEach((k, v) -> {...});` is a better shorthand in cases like these here where you don't need to pass the entry between stream stages.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java:
```diff
@@ -97,12 +103,65 @@ public void shouldNotCreateTopicWhenItAlreadyExists() {
+                }
+
+                // Try clusters with different default replication factors
+                for (int defaultRF = 1; defaultRF != maxDefaultRf; ++defaultRF) {
```
I didn't comment elsewhere, but that's where `!=` has the potential of stackoverflow vs `<=`. I understand we trust ourselves to set `maxDefaultRf` to be `>0` but that's a couple indirections away.

kkonstantine reviewed (commented):
Thanks for the PR @rhauch 
This looks pretty good already! 

I left a few minor comments after a first pass.

kkonstantine commented:
fyi, builds fail on checkstyle.

rhauch committed:
KAFKA-9931: Incorporated suggestions from review

rhauch committed:
KAFKA-9931: Fix checkstyle failures

rhauch commented:
Thanks for the review, @kkonstantine. I think I've incorporated all of your suggestions.

rhauch commented:
ok to test

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -400,6 +424,33 @@ public KeyGenerator getInternalRequestKeyGenerator() {
+        if (CONFIG_STORAGE_PREFIX.equals(prefix) && result.containsKey(PARTITIONS_SUFFIX)) {
+            log.warn("Ignoring '{}{}={}' setting, since config topic partitions is always 1", prefix, PARTITIONS_SUFFIX, result.get("partitions"));
+        }
+        Object removedPolicy = result.remove("cleanup.policy");
```
It's one more import statement, but I don't see a reason not to get the actual config from: 
`TopicConfig.CLEANUP_POLICY_CONFIG`. I think you'll agree that a variable is better than a string, but just a nit.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java:
```diff
@@ -453,11 +453,17 @@ public void putSessionKey(SessionKey sessionKey) {
-                replicationFactor(config.getShort(DistributedConfig.CONFIG_STORAGE_REPLICATION_FACTOR_CONFIG)).
-                build();
+
+        Map<String, Object> topicSettings = null;
+        if (config instanceof DistributedConfig) {
+            topicSettings = ((DistributedConfig) config).configStorageTopicSettings();
+        }
```
```suggestion
        Map<String, Object> topicSettings = config instanceof DistributedConfig
                ? ((DistributedConfig) config).configStorageTopicSettings()
                : Collections.emptyMap();
```
I know that `TopicAdmin#defineTopic` checks for `null`, but I think using `null` with collections is better to do when such optimization matters. Wdyt? 
(btw you don't have to use the ternary operator, I just added it to make the suggestion clear). 

Also, if you change here, please change in the other files too.

kkonstantine reviewed (approved):
I think it's ready. 
Added a few optional comments that caught my eye on the last pass, but otherwise
LGTM

kkonstantine commented:
retest this please

rhauch committed:
KAFKA-9931: Used existing constants in TopicAdmin

rhauch committed:
KAFKA-9931: Applied recommendations from code reviews

rhauch committed:
KAFKA-9931: Improved comments

rhauch commented:
Incorporated @kkonstantine's more recent suggestions, and further changed `TopicAdmin` to use existing `TopicConfig` constants rather than string literals. 

Rebased to pick up the changes from #8653 rather than incorporating the same commit in this PR.

rhauch commented:
FYI: previous builds had no failures related to Connect.

kkonstantine reviewed (approved):
Thanks @rhauch 
LGTM, expecting that the current build won't have related failures as well.

rhauch commented:
retest this please

rhauch commented:
All Connect unit and integration tests passed in all of the builds, but each of the builds failed due to a few flaky integration tests in either Core or Streams.
